lockfile/src/sbom.rs

```diff
@@ -0,0 +1,222 @@
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Clone)]
+#[serde(rename_all = "SCREAMING-KEBAB-CASE")]
+pub enum ReferenceCategory {
+    Other,
+    // older schema uses _
+    #[serde(alias = "PERSISTENT_ID")]
+    PersistentId,
+    Security,
+    // older schema uses _
+    #[serde(alias = "PACKAGE_MANAGER")]
+    PackageManager,
+}
+
+impl TryFrom<&PackageInformation> for Package {
+    type Error = anyhow::Error;
+
+    fn try_from(pkg_info: &PackageInformation) -> anyhow::Result<Self> {
+        let pkg_url = pkg_info
+            .external_refs
+            .iter()
+            .find_map(|external| match external.reference_category {
+                ReferenceCategory::PackageManager => {
+                    if external.reference_type == "purl" {
+                        Some(&external.reference_locator)
+                    } else {
+                        None
+                    }
+                },
+                _ => None,
+            })
+            .context("Package manager not found")?;
+
+        let purl = PackageUrl::from_str(pkg_url).context("Unable to parse package url")?;
+        let package_type =
+            PackageType::from_str(purl.ty()).map_err(|_| anyhow!("Unrecognized ecosystem"))?;
+        let name = match purl.namespace() {
+            Some(ns) => format!("{}/{}", ns, purl.name()),
+            None => purl.name().into(),
+        };
+        let pkg_version = match (&pkg_info.version_info, purl.version()) {
+            (Some(v), _) => Some(v.to_string()),
+            (None, Some(v)) => Some(v.into()),
+            _ => None,
+        }
+        .context("Unable to determine version")?;
+
+        let version = purl
+            .qualifiers()
+            .iter()
+            .find_map(|(key, value)| match key.as_ref() {
+                "repository_url" => Some(PackageVersion::ThirdParty(ThirdPartyVersion {
+                    version: pkg_version.clone(),
+                    registry: value.to_string(),
+                })),
+                "download_url" => Some(PackageVersion::DownloadUrl(value.to_string())),
+                "vcs_url" => {
+                    if value.as_ref().starts_with("git+") {
+                        Some(PackageVersion::Git(value.to_string()))
+                    } else {
+                        None
+                    }
+                },
+                _ => None,
+            })
+            .unwrap_or(PackageVersion::FirstParty(pkg_version));
+
+        Ok(Package { name, version, package_type })
+    }
+}
+
+pub struct Sbom;
+
+impl Parse for Sbom {
+    fn parse(&self, data: &str) -> anyhow::Result<Vec<Package>> {
+        let mut lock: Spdx = serde_json::from_str(data).unwrap_or(serde_yaml::from_str(data)?);
+
+        let packages = lock
+            .packages
+            .drain(..)
+            .filter_map(|package_info| Package::try_from(&package_info).ok())
+            .collect::<Vec<Package>>();
```

Comment:
Silently ignoring packages has led to issues in the past (and ultimately led to #882).

Errors in `Package::try_from` should not be ignored.

```suggestion
        let packages = Vec::new();
        for package_info in lock.packages {
            packages.push(Package::try_from(&package_info)?);
        }
```

Reply:
There are some issues with allowing all packages.

1. Package `versionInfo` is optional. If it's provided in the external refs, then that's used.
2. There isn't a good way to determine which ecosystem the package comes from without a `PURL`. Although I can add some logic to look at the `downloadLocation`, which is always present but can be `NOASSERTION`, and compare URLs with first party registries. Without the `PURL` and maybe `downloadLocation`, there's no good way to determine the ecosystem.
3. Sboms will contain ecosystems we currently do not support. This will prevent anyone from submitting an sbom and analyzing packages from ecosystems we do support.

I built a container targeting all the ecosystems we currently support and scanned it with [syft](https://github.com/anchore/syft) to generate some of the test files. Guaranteed we'll be ingesting sboms generated by syft.


Ex1: This has a `PURL` so we can determine it's maven package but there's no version info.
```
  {
   "name": "ca-certificates-java",
   "SPDXID": "SPDXRef-Package-java-archive-ca-certificates-java-300cbdd37d289bad",
   "downloadLocation": "NOASSERTION",
   "checksums": [
    {
     "algorithm": "SHA1",
     "checksumValue": "0b1d4f423b4ec416cb14026a75c184bdb06c6914"
    }
   ],
   "sourceInfo": "acquired package info from installed java archive: /usr/share/ca-certificates-java/ca-certificates-java.jar",
   "licenseConcluded": "NONE",
   "licenseDeclared": "NONE",
   "copyrightText": "NOASSERTION",
   "externalRefs": [
    {
     "referenceCategory": "PACKAGE-MANAGER",
     "referenceType": "purl",
     "referenceLocator": "pkg:maven/org.debian.security.UpdateCertificates/ca-certificates-java"
    }
   ]
  },
  ```

Ex2: No `PURL` but there is a `downloadLocation`. It's also something we don't support at the moment.
```
- SPDXID: "SPDXRef-Package"
  annotations:
  - annotationDate: "2011-01-29T18:30:22Z"
    annotationType: "OTHER"
    annotator: "Person: Package Commenter"
    comment: "Package level annotation"
  attributionTexts:
  - "The GNU C Library is free software.  See the file COPYING.LIB for copying conditions,\
    \ and LICENSES for notices about a few contributions that require these additional\
    \ notices to be distributed.  License copyright years may be listed using range\
    \ notation, e.g., 1996-2015, indicating that every year in the range, inclusive,\
    \ is a copyrightable year that would otherwise be listed individually."
  builtDate: "2011-01-29T18:30:22Z"
  checksums:
  - algorithm: "MD5"
    checksumValue: "624c1abb3664f4b35547e7c73864ad24"
  - algorithm: "SHA1"
    checksumValue: "85ed0817af83a24ad8da68c2b5094de69833983c"
  - algorithm: "SHA256"
    checksumValue: "11b6d3ee554eedf79299905a98f9b9a04e498210b59f15094c916c91d150efcd"
  - algorithm: "BLAKE2b-384"
    checksumValue: "aaabd89c926ab525c242e6621f2f5fa73aa4afe3d9e24aed727faaadd6af38b620bdb623dd2b4788b1c8086984af8706"
  copyrightText: "Copyright 2008-2010 John Smith"
  description: "The GNU C Library defines functions that are specified by the ISO\
    \ C standard, as well as additional features specific to POSIX and other derivatives\
    \ of the Unix operating system, and extensions specific to GNU systems."
  downloadLocation: "http://ftp.gnu.org/gnu/glibc/glibc-ports-2.15.tar.gz"
  externalRefs:
  - referenceCategory: "SECURITY"
    referenceLocator: "cpe:2.3:a:pivotal_software:spring_framework:4.1.0:*:*:*:*:*:*:*"
    referenceType: "cpe23Type"
  - comment: "This is the external ref for Acme"
    referenceCategory: "OTHER"
    referenceLocator: "acmecorp/acmenator/4.1.3-alpha"
    referenceType: "http://spdx.org/spdxdocs/spdx-example-444504E0-4F89-41D3-9A0C-0305E82C3301#LocationRef-acmeforge"
  filesAnalyzed: true
  homepage: "http://ftp.gnu.org/gnu/glibc"
  licenseComments: "The license for this project changed with the release of version\
    \ x.y.  The version of the project included here post-dates the license change."
  licenseConcluded: "(LGPL-2.0-only OR LicenseRef-3)"
  licenseDeclared: "(LGPL-2.0-only AND LicenseRef-3)"
  licenseInfoFromFiles:
  - "GPL-2.0-only"
  - "LicenseRef-2"
  - "LicenseRef-1"
  name: "glibc"
  originator: "Organization: ExampleCodeInspect (contact@example.com)"
  packageFileName: "glibc-2.11.1.tar.gz"
  packageVerificationCode:
    packageVerificationCodeExcludedFiles:
    - "./package.spdx"
    packageVerificationCodeValue: "d6a770ba38583ed4bb4525bd96e50461655d2758"
  primaryPackagePurpose: "SOURCE"
  hasFiles:
  - "SPDXRef-Specification"
  - "SPDXRef-Specification"
  - "SPDXRef-CommonsLangSrc"
  - "SPDXRef-Specification"
  - "SPDXRef-CommonsLangSrc"
  - "SPDXRef-JenaLib"
  - "SPDXRef-Specification"
  - "SPDXRef-CommonsLangSrc"
  - "SPDXRef-JenaLib"
  - "SPDXRef-DoapSource"
  - "SPDXRef-Specification"
  - "SPDXRef-CommonsLangSrc"
  - "SPDXRef-JenaLib"
  - "SPDXRef-DoapSource"
  releaseDate: "2012-01-29T18:30:22Z"
  sourceInfo: "uses glibc-2_11-branch from git://sourceware.org/git/glibc.git."
  summary: "GNU C library."
  supplier: "Person: Jane Doe (jane.doe@example.com)"
  validUntilDate: "2014-01-29T18:30:22Z"
  versionInfo: "2.11.1"
  ```

I guess I'm looking for direction on what should we do with regards to missing versions, unable to determine ecosystems (do we add an "unknown"), and for ecosystems we don't support.

Reply:
I'll take the easy case first.

> * Sboms will contain ecosystems we currently do not support. This will prevent anyone from submitting an sbom and analyzing packages from ecosystems we do support.

This should be a **very loud** warning. Users need to be aware that we do not yet support these ecosystems.

> * Package `versionInfo` is optional. If it's provided in the external refs, then that's used.

If there is no indication of package version, I think it's fair to reject the SBOM. We (currently) do not even support version ranges. And **no** version at all is even worse. This should probably be a hard error.

> * There isn't a good way to determine which ecosystem the package comes from without a `PURL`. Although I can add some logic to look at the `downloadLocation`, which is always present but can be `NOASSERTION`, and compare URLs with first party registries. Without the `PURL` and maybe `downloadLocation`, there's no good way to determine the ecosystem.

This is a bit more of a gray area... But I also think this should be a hard error for now. If PURLs are required for us to parse an SBOM, then we should reject SBOMs that don't have PURLs.

I know it seems harsh to reject entire files for errors like these, but it's better than giving the green light to a file that we barely understand. Our SBOM support will evolve and improve with time.

Reply:
The thing about SBOMs is that _most_ of the stuff in them is optional. It's kind of a useless file imo. So I think loud warnings and hard errors are entirely appropriate here.